cmd/dep/status.go

```diff
@@ -1068,6 +1068,33 @@ func collectConstraints(ctx *dep.Ctx, p *dep.Project, sm gps.SourceManager) (con
 		}
 	}
 
+	// Incorporate constraints set in the manifest of the root project.
+	if p.Manifest != nil {
+
+		ineffCon := p.FindIneffectualConstraints(sm)
+
+		// Iterate through regular project constraints, append if it is not
+		// an ineffectual constraint
+		for pr, pp := range p.Manifest.Constraints {
+			i := sort.Search(len(ineffCon), func(i int) bool {
+				return pr == ineffCon[i]
+			})
```

Comment:
I'm a little confused about this. I would need a little help with the usage of binary search here. Maybe a little explanation?

Reply:
FindIneffectualConstraints returns a _sorted_ list of all the ineffective project roots. For every project root defined in the constraints of the projects manifest, we need to make sure it is not an ineffectual constraint. Instead of converting the result from FindIneffectualConstraints into a map or iterating over the whole range of its result, we utilize the property that it is sorted.

Please lmk if you want me to add the explanation to the code as well or if you prefer a different approach altogether.

Reply:
We don't need to use `FindIneffectualConstraints`. We can just use `directDeps` to check if the project root (pr) of the constraint is a direct dependency of the main root project. If yes, add it in the `constraintCollection`, else check the next project in manifest. Also, since this is collecting the constraints set by the main root project, it should be named as `"root"`. So that would look something like this:
```golang
if _, ok := directDeps[pr]; !ok {
    continue
}

tempCC := append(
    constraintCollection[string(pr)],
    projectConstraint{"root", pp.Constraint},
)
```
Sorry I couldn't come up with all these last week. A little out of touch, took some time to recall 😅